<!doctype html>
<html lang="zh-CN">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width">
  <title>错误的冒泡和捕获 | Node.js API 文档</title>
  <link rel="icon" sizes="32x32" type="image/png" href="http://static.nodejs.cn/_static/img/favicon.png"><link rel="stylesheet" href="/api/api.css">
  <script>
    document.write('<link href="http://nodejs.cn/api/errors/'&#32;+&#32;(location.port?'':'//static.nodejs.cn')&#32;+&#32;'/_static/css/api.css?1576756305125" rel="stylesheet">')
  </script>
</head>
<body class="alt apidoc" id="page_api_item">
  <div id="api-section-errors">
    <div id="content" class="clearfix">
  
      <div id="column1" data-id="__ID__" class="interior">
        <header>
          <h1>错误的冒泡和捕获</h1>
          <div id="gtoc">
            <ul>
              <li>v12.14.0</li>
              <li>
                <a href="../errors.html#errors_error_propagation_and_interception" name="toc">返回上层文档</a>
              </li>
              <li>
                <a href="https://github.com/nodejscn/node-api-cn/edit/master/errors/error_propagation_and_interception.md" rel="nofollow" target="_blank">提交修改</a>
              </li>
              <li>
                <a href="http://nodejs.cn/search" name="toc">搜索</a>
              </li>
            </ul>
          </div>
          <hr>
        </header>

        <div id="biz_item"></div>
  
        <div id="apicontent">
          <div id="content_left"><p>Node.js &#x652F;&#x6301;&#x51E0;&#x79CD;&#x5F53;&#x5E94;&#x7528;&#x7A0B;&#x5E8F;&#x8FD0;&#x884C;&#x65F6;&#x53D1;&#x751F;&#x7684;&#x9519;&#x8BEF;&#x7684;&#x5192;&#x6CE1;&#x548C;&#x5904;&#x7406;&#x7684;&#x673A;&#x5236;&#x3002;
&#x5982;&#x4F55;&#x62A5;&#x544A;&#x548C;&#x5904;&#x7406;&#x8FD9;&#x4E9B;&#x9519;&#x8BEF;&#x5B8C;&#x5168;&#x53D6;&#x51B3;&#x4E8E; <code>Error</code> &#x7684;&#x7C7B;&#x578B;&#x548C;&#x88AB;&#x8C03;&#x7528;&#x7684; API &#x7684;&#x98CE;&#x683C;&#x3002;</p><p>&#x6240;&#x6709; JavaScript &#x9519;&#x8BEF;&#x90FD;&#x4F1A;&#x88AB;&#x4F5C;&#x4E3A;&#x5F02;&#x5E38;&#x5904;&#x7406;&#xFF0C;&#x5F02;&#x5E38;&#x4F1A;&#x7ACB;&#x5373;&#x4EA7;&#x751F;&#x5E76;&#x4F7F;&#x7528;&#x6807;&#x51C6;&#x7684; JavaScript <code>throw</code> &#x673A;&#x5236;&#x629B;&#x51FA;&#x4E00;&#x4E2A;&#x9519;&#x8BEF;&#x3002;
&#x8FD9;&#x4E9B;&#x90FD;&#x662F;&#x4F7F;&#x7528; JavaScript &#x8BED;&#x8A00;&#x63D0;&#x4F9B;&#x7684; <a href="http://nodejs.cn/s/7W62Fd" rel="nofollow"><code>try&#x2026;catch</code> &#x8BED;&#x53E5;</a>&#x5904;&#x7406;&#x7684;&#x3002;</p><pre><code class="language-js">// &#x629B;&#x51FA;&#x4E00;&#x4E2A; ReferenceError&#xFF0C;&#x56E0;&#x4E3A; z &#x672A;&#x5B9A;&#x4E49;&#x3002;
try {
  const m = 1;
  const n = m + z;
} catch (err) {
  // &#x5728;&#x8FD9;&#x91CC;&#x5904;&#x7406;&#x9519;&#x8BEF;&#x3002;
}
</code></pre><p>JavaScript &#x7684; <code>throw</code> &#x673A;&#x5236;&#x7684;&#x4EFB;&#x4F55;&#x4F7F;&#x7528;&#x90FD;&#x4F1A;&#x5F15;&#x8D77;&#x5F02;&#x5E38;&#xFF0C;&#x5F02;&#x5E38;&#x5FC5;&#x987B;&#x4F7F;&#x7528; <code>try&#x2026;catch</code> &#x5904;&#x7406;&#xFF0C;&#x5426;&#x5219; Node.js &#x8FDB;&#x7A0B;&#x4F1A;&#x7ACB;&#x5373;&#x9000;&#x51FA;&#x3002;</p><p>&#x9664;&#x4E86;&#x5C11;&#x6570;&#x4F8B;&#x5916;&#xFF0C;&#x540C;&#x6B65;&#x7684; API&#xFF08;&#x4EFB;&#x4F55;&#x4E0D;&#x63A5;&#x53D7; <code>callback</code> &#x51FD;&#x6570;&#x7684;&#x963B;&#x585E;&#x65B9;&#x6CD5;&#xFF0C;&#x4F8B;&#x5982; <a href="http://nodejs.cn/s/LxKo1q" rel="nofollow"><code>fs.readFileSync</code></a>&#xFF09;&#x4F1A;&#x4F7F;&#x7528; <code>throw</code> &#x62A5;&#x544A;&#x9519;&#x8BEF;&#x3002;</p><p>&#x5F02;&#x6B65;&#x7684; API &#x4E2D;&#x53D1;&#x751F;&#x7684;&#x9519;&#x8BEF;&#x53EF;&#x80FD;&#x4F1A;&#x4EE5;&#x591A;&#x79CD;&#x65B9;&#x5F0F;&#x8FDB;&#x884C;&#x62A5;&#x544A;&#xFF1A;</p><ul>
<li>&#x5927;&#x591A;&#x6570;&#x7684;&#x5F02;&#x6B65;&#x65B9;&#x6CD5;&#x90FD;&#x63A5;&#x53D7;&#x4E00;&#x4E2A; <code>callback</code> &#x51FD;&#x6570;&#xFF0C;&#x8BE5;&#x51FD;&#x6570;&#x4F1A;&#x63A5;&#x53D7;&#x4E00;&#x4E2A; <code>Error</code> &#x5BF9;&#x8C61;&#x4F20;&#x5165;&#x4F5C;&#x4E3A;&#x7B2C;&#x4E00;&#x4E2A;&#x53C2;&#x6570;&#x3002;
&#x5982;&#x679C;&#x7B2C;&#x4E00;&#x4E2A;&#x53C2;&#x6570;&#x4E0D;&#x662F; <code>null</code> &#x800C;&#x662F;&#x4E00;&#x4E2A; <code>Error</code> &#x5B9E;&#x4F8B;&#xFF0C;&#x5219;&#x8BF4;&#x660E;&#x53D1;&#x751F;&#x4E86;&#x9519;&#x8BEF;&#xFF0C;&#x5E94;&#x8BE5;&#x8FDB;&#x884C;&#x5904;&#x7406;&#x3002;</li>
</ul><pre><code class="language-js">const fs = require(&apos;fs&apos;);
fs.readFile(&apos;&#x4E00;&#x4E2A;&#x4E0D;&#x5B58;&#x5728;&#x7684;&#x6587;&#x4EF6;&apos;, (err, data) =&gt; {
  if (err) {
    console.error(&apos;&#x8BFB;&#x53D6;&#x6587;&#x4EF6;&#x51FA;&#x9519;&#xFF01;&apos;, err);
    return;
  }
  // &#x5426;&#x5219;&#x5904;&#x7406;&#x6570;&#x636E;
});
</code></pre><ul>
<li>
<p>&#x5F53;&#x4E00;&#x4E2A;&#x5F02;&#x6B65;&#x65B9;&#x6CD5;&#x88AB;&#x4E00;&#x4E2A; <a href="http://nodejs.cn/s/pGAddE" rel="nofollow"><code>EventEmitter</code></a> &#x5BF9;&#x8C61;&#x8C03;&#x7528;&#x65F6;&#xFF0C;&#x9519;&#x8BEF;&#x4F1A;&#x88AB;&#x5206;&#x53D1;&#x5230;&#x5BF9;&#x8C61;&#x7684; <code>&apos;error&apos;</code> &#x4E8B;&#x4EF6;&#x4E0A;&#x3002;</p>
<pre><code class="language-js">const net = require(&apos;net&apos;);
const connection = net.connect(&apos;localhost&apos;);

// &#x6DFB;&#x52A0;&#x4E00;&#x4E2A; &apos;error&apos; &#x4E8B;&#x4EF6;&#x53E5;&#x67C4;&#x5230;&#x4E00;&#x4E2A;&#x6D41;&#xFF1A;
connection.on(&apos;error&apos;, (err) =&gt; {
  // &#x5982;&#x679C;&#x8FDE;&#x63A5;&#x88AB;&#x670D;&#x52A1;&#x5668;&#x91CD;&#x7F6E;&#xFF0C;&#x6216;&#x65E0;&#x6CD5;&#x8FDE;&#x63A5;&#xFF0C;&#x6216;&#x53D1;&#x751F;&#x4EFB;&#x4F55;&#x9519;&#x8BEF;&#xFF0C;&#x5219;&#x9519;&#x8BEF;&#x4F1A;&#x88AB;&#x53D1;&#x9001;&#x5230;&#x8FD9;&#x91CC;&#x3002; 
  console.error(err);
});

connection.pipe(process.stdout);
</code></pre>
</li>
<li>
<p>Node.js API &#x4E2D;&#x6709;&#x4E00;&#x5C0F;&#x90E8;&#x5206;&#x666E;&#x901A;&#x7684;&#x5F02;&#x6B65;&#x65B9;&#x6CD5;&#x4ECD;&#x53EF;&#x80FD;&#x4F7F;&#x7528; <code>throw</code> &#x673A;&#x5236;&#x629B;&#x51FA;&#x5F02;&#x5E38;&#xFF0C;&#x4E14;&#x5FC5;&#x987B;&#x4F7F;&#x7528; <code>try&#x2026;catch</code> &#x5904;&#x7406;&#x3002;
&#x8FD9;&#x4E9B;&#x65B9;&#x6CD5;&#x5E76;&#x6CA1;&#x6709;&#x4E00;&#x4E2A;&#x5B8C;&#x6574;&#x7684;&#x5217;&#x8868;&#xFF1B;&#x8BF7;&#x53C2;&#x9605;&#x5404;&#x4E2A;&#x65B9;&#x6CD5;&#x7684;&#x6587;&#x6863;&#x4EE5;&#x786E;&#x5B9A;&#x6240;&#x9700;&#x7684;&#x5408;&#x9002;&#x7684;&#x9519;&#x8BEF;&#x5904;&#x7406;&#x673A;&#x5236;&#x3002;</p>
</li>
</ul><p><code>&apos;error&apos;</code> &#x4E8B;&#x4EF6;&#x673A;&#x5236;&#x7684;&#x4F7F;&#x7528;&#x5E38;&#x89C1;&#x4E8E;<a href="http://nodejs.cn/s/kUvpNm" rel="nofollow">&#x57FA;&#x4E8E;&#x6D41;</a>&#x548C;<a href="http://nodejs.cn/s/pGAddE" rel="nofollow">&#x57FA;&#x4E8E;&#x4E8B;&#x4EF6;&#x89E6;&#x53D1;&#x5668;</a>&#x7684; API&#xFF0C;&#x5B83;&#x4EEC;&#x672C;&#x8EAB;&#x5C31;&#x4EE3;&#x8868;&#x4E86;&#x4E00;&#x7CFB;&#x5217;&#x7684;&#x5F02;&#x6B65;&#x64CD;&#x4F5C;&#xFF08;&#x76F8;&#x5BF9;&#x4E8E;&#x8981;&#x4E48;&#x6210;&#x529F;&#x8981;&#x4E48;&#x5931;&#x8D25;&#x7684;&#x5355;&#x4E00;&#x64CD;&#x4F5C;&#xFF09;&#x3002;</p><p>&#x5BF9;&#x4E8E;&#x6240;&#x6709;&#x7684; <a href="http://nodejs.cn/s/pGAddE" rel="nofollow"><code>EventEmitter</code></a> &#x5BF9;&#x8C61;&#xFF0C;&#x5982;&#x679C;&#x6CA1;&#x6709;&#x63D0;&#x4F9B;&#x4E00;&#x4E2A; <code>&apos;error&apos;</code> &#x4E8B;&#x4EF6;&#x53E5;&#x67C4;&#xFF0C;&#x5219;&#x9519;&#x8BEF;&#x4F1A;&#x88AB;&#x629B;&#x51FA;&#xFF0C;&#x5E76;&#x9020;&#x6210; Node.js &#x8FDB;&#x7A0B;&#x62A5;&#x544A;&#x4E00;&#x4E2A;&#x672A;&#x5904;&#x7406;&#x7684;&#x5F02;&#x5E38;&#x4E14;&#x968F;&#x5373;&#x5D29;&#x6E83;&#xFF0C;&#x9664;&#x975E;&#xFF1A;
&#x9002;&#x5F53;&#x5730;&#x4F7F;&#x7528; <a href="http://nodejs.cn/s/cnfQ9s" rel="nofollow"><code>domain</code></a> &#x6A21;&#x5757;&#x6216;&#x5DF2;&#x7ECF;&#x6CE8;&#x518C;&#x4E86;&#x4E00;&#x4E2A; <a href="http://nodejs.cn/s/Lc6A38" rel="nofollow"><code>&apos;uncaughtException&apos;</code></a> &#x4E8B;&#x4EF6;&#x7684;&#x53E5;&#x67C4;&#x3002;</p><pre><code class="language-js">const EventEmitter = require(&apos;events&apos;);
const ee = new EventEmitter();

setImmediate(() =&gt; {
  // &#x8FD9;&#x4F1A;&#x4F7F;&#x8FDB;&#x7A0B;&#x5D29;&#x6E83;&#xFF0C;&#x56E0;&#x4E3A;&#x8FD8;&#x4E3A;&#x6DFB;&#x52A0; &apos;error&apos; &#x4E8B;&#x4EF6;&#x53E5;&#x67C4;&#x3002;
  ee.emit(&apos;error&apos;, new Error(&apos;&#x8FD9;&#x4F1A;&#x5D29;&#x6E83;&apos;));
});
</code></pre><p>&#x8FD9;&#x79CD;&#x65B9;&#x5F0F;&#x4EA7;&#x751F;&#x7684;&#x9519;&#x8BEF;&#x65E0;&#x6CD5;&#x4F7F;&#x7528; <code>try&#x2026;catch</code> &#x622A;&#x83B7;&#xFF0C;&#x56E0;&#x4E3A;&#x5B83;&#x4EEC;&#x662F;&#x5728;&#x8C03;&#x7528;&#x7684;&#x4EE3;&#x7801;&#x5DF2;&#x7ECF;&#x9000;&#x51FA;&#x540E;&#x629B;&#x51FA;&#x7684;&#x3002;</p><p>&#x5F00;&#x53D1;&#x8005;&#x5FC5;&#x987B;&#x67E5;&#x9605;&#x5404;&#x4E2A;&#x65B9;&#x6CD5;&#x7684;&#x6587;&#x6863;&#x4EE5;&#x660E;&#x786E;&#x5728;&#x9519;&#x8BEF;&#x53D1;&#x751F;&#x65F6;&#x8FD9;&#x4E9B;&#x65B9;&#x6CD5;&#x662F;&#x5982;&#x4F55;&#x5192;&#x6CE1;&#x7684;&#x3002;</p></div>
          <div id="content_right"><p>Node.js supports several mechanisms for propagating and handling errors that
occur while an application is running. How these errors are reported and
handled depends entirely on the type of <code>Error</code> and the style of the API that is
called.</p><p>All JavaScript errors are handled as exceptions that <em>immediately</em> generate
and throw an error using the standard JavaScript <code>throw</code> mechanism. These
are handled using the <a href="http://nodejs.cn/s/7W62Fd" rel="nofollow"><code>try&#x2026;catch</code> construct</a> provided by the
JavaScript language.</p><pre><code class="language-js">// Throws with a ReferenceError because z is not defined.
try {
  const m = 1;
  const n = m + z;
} catch (err) {
  // Handle the error here.
}
</code></pre><p>Any use of the JavaScript <code>throw</code> mechanism will raise an exception that
<em>must</em> be handled using <code>try&#x2026;catch</code> or the Node.js process will exit
immediately.</p><p>With few exceptions, <em>Synchronous</em> APIs (any blocking method that does not
accept a <code>callback</code> function, such as <a href="http://nodejs.cn/s/LxKo1q" rel="nofollow"><code>fs.readFileSync</code></a>), will use <code>throw</code>
to report errors.</p><p>Errors that occur within <em>Asynchronous APIs</em> may be reported in multiple ways:</p><ul>
<li>Most asynchronous methods that accept a <code>callback</code> function will accept an
<code>Error</code> object passed as the first argument to that function. If that first
argument is not <code>null</code> and is an instance of <code>Error</code>, then an error occurred
that should be handled.</li>
</ul><pre><code class="language-js">const fs = require(&apos;fs&apos;);
fs.readFile(&apos;a file that does not exist&apos;, (err, data) =&gt; {
  if (err) {
    console.error(&apos;There was an error reading the file!&apos;, err);
    return;
  }
  // Otherwise handle the data
});
</code></pre><ul>
<li>
<p>When an asynchronous method is called on an object that is an
<a href="http://nodejs.cn/s/pGAddE" rel="nofollow"><code>EventEmitter</code></a>, errors can be routed to that object&apos;s <code>&apos;error&apos;</code> event.</p>
<pre><code class="language-js">const net = require(&apos;net&apos;);
const connection = net.connect(&apos;localhost&apos;);

// Adding an &apos;error&apos; event handler to a stream:
connection.on(&apos;error&apos;, (err) =&gt; {
  // If the connection is reset by the server, or if it can&apos;t
  // connect at all, or on any sort of error encountered by
  // the connection, the error will be sent here.
  console.error(err);
});

connection.pipe(process.stdout);
</code></pre>
</li>
<li>
<p>A handful of typically asynchronous methods in the Node.js API may still
use the <code>throw</code> mechanism to raise exceptions that must be handled using
<code>try&#x2026;catch</code>. There is no comprehensive list of such methods; please
refer to the documentation of each method to determine the appropriate
error handling mechanism required.</p>
</li>
</ul><p>The use of the <code>&apos;error&apos;</code> event mechanism is most common for <a href="http://nodejs.cn/s/kUvpNm" rel="nofollow">stream-based</a>
and <a href="http://nodejs.cn/s/pGAddE" rel="nofollow">event emitter-based</a> APIs, which themselves represent a series of
asynchronous operations over time (as opposed to a single operation that may
pass or fail).</p><p>For <em>all</em> <a href="http://nodejs.cn/s/pGAddE" rel="nofollow"><code>EventEmitter</code></a> objects, if an <code>&apos;error&apos;</code> event handler is not
provided, the error will be thrown, causing the Node.js process to report an
uncaught exception and crash unless either: The <a href="http://nodejs.cn/s/cnfQ9s" rel="nofollow"><code>domain</code></a> module is
used appropriately or a handler has been registered for the
<a href="http://nodejs.cn/s/Lc6A38" rel="nofollow"><code>&apos;uncaughtException&apos;</code></a> event.</p><pre><code class="language-js">const EventEmitter = require(&apos;events&apos;);
const ee = new EventEmitter();

setImmediate(() =&gt; {
  // This will crash the process because no &apos;error&apos; event
  // handler has been added.
  ee.emit(&apos;error&apos;, new Error(&apos;This will crash&apos;));
});
</code></pre><p>Errors generated in this way <em>cannot</em> be intercepted using <code>try&#x2026;catch</code> as
they are thrown <em>after</em> the calling code has already exited.</p><p>Developers must refer to the documentation for each method to determine
exactly how errors raised by those methods are propagated.</p></div>
          <div></div>
        </div>
      </div>
    </div>
  </div>
  <script>
    document.write('<script src="http://nodejs.cn/api/errors/'&#32;+&#32;(location.port&#32;?&#32;''&#32;:&#32;'//static.nodejs.cn')&#32;+&#32;'/_static/js/api.js?1576756305125"><\/script>');
  </script>
</body>
</html>